### PR TITLE
add matching lending order

### PR DESCRIPTION
**Pull request by nguyenbatam**

### CODE REVIEW

File: common/constants.go [hunk 1]
```diff
@@ -38,9 +38,11 @@ var BasePrice = big.NewInt(1000000000000000000) // 1
...
 var RelayerFee = big.NewInt(1000000000000000)   // 0.001
 var TomoXBaseFee = big.NewInt(10000)
-var RelayerCancelFee = big.NewInt(100000000000000) // 0.0001
-var TomoXBaseCancelFee = big.NewInt(10000)         // 1/10000
-
+var RelayerCancelFee = big.NewInt(100000000000000)    // 0.0001
+var TomoXBaseCancelFee = big.NewInt(10000)            // 1/10000
+var RelayerLendingFee = big.NewInt(10000000000000000) // 0.01
+var LendingCancelFee = big.NewInt(1000000)            // 1/1000000
+var BaseLendingInterest = big.NewInt(1000)
```
Comment (thanhnguyennguyen) [inline common/constants.go]: should be 10 000

File: common/types.go [hunk 2]
```diff
@@ -38,6 +38,7 @@ const (
 	TomoXAddr           = "0x0000000000000000000000000000000000000091"
 	TomoXStateAddr      = "0x0000000000000000000000000000000000000092"
 	TomoNativeAddress   = "0x0000000000000000000000000000000000000001"
+	LendingLockAddress  = "0x0000000000000000000000000000000000000001"
```
Comment (thanhnguyennguyen) [inline common/types.go]: use another address to avoid conflicting with `TomoNativeAddress`

File: tomoxlending/order_processor.go [hunk 3]
```diff
@@ -238,25 +237,45 @@ func (l *Lending) processOrderList(createdBlockTime uint64,coinbase common.Addre
...
+		}
+		collateralPrice := common.BasePrice
+		depositRate, liquidationRate, collateralTOMOBasePrice := lendingstate.GetCollateralDetail(statedb, collateralToken)
+
+		_, _, lendingTOMOBasePrice := lendingstate.GetCollateralDetail(statedb, order.LendingToken)
+		lendTokenTOMOPrice, err := l.GetMediumPriceLastEpoch(chain, statedb, tradingStateDb, common.HexToAddress(common.TomoNativeAddress), order.LendingToken)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		if lendTokenTOMOPrice == nil || lendTokenTOMOPrice.Sign() == 0 {
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go]: lendingToken price should be got from tomox.

File: tomoxlending/order_processor.go [hunk 4]
```diff
@@ -238,25 +237,45 @@ func (l *Lending) processOrderList(createdBlockTime uint64,coinbase common.Addre
...
+		collateralToken := order.CollateralToken
+		borrowFee := lendingstate.GetFee(statedb, order.Relayer)
+		if order.Side == lendingstate.Investing {
+			collateralToken = oldestOrder.CollateralToken
+			borrowFee = lendingstate.GetFee(statedb, oldestOrder.Relayer)
+		}
+		collateralPrice := common.BasePrice
+		depositRate, liquidationRate, collateralTOMOBasePrice := lendingstate.GetCollateralDetail(statedb, collateralToken)
+
+		_, _, lendingTOMOBasePrice := lendingstate.GetCollateralDetail(statedb, order.LendingToken)
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go]: the lending contract provides the price of collaterals only, doesn't have the price of lending token.
Price of lending token (must be available for trading in tomox) should be got from tomox.

File: tomoxlending/order_processor.go [hunk 5]
```diff
@@ -400,203 +440,195 @@ func (l *Lending) getTradeQuantity(quoteInterest *big.Int, coinbase common.Addre
...
-func DoSettleBalance(coinbase common.Address, takerOrder, makerOrder *lendingstate.LendingItem, settleBalance *lendingstate.SettleBalance, statedb *state.StateDB) error {
+func DoSettleBalance(coinbase common.Address, takerOrder, makerOrder *lendingstate.LendingItem, settleBalance *lendingstate.LendingSettleBalance, statedb *state.StateDB) error {
 	takerExOwner := lendingstate.GetRelayerOwner(takerOrder.Relayer, statedb)
 	makerExOwner := lendingstate.GetRelayerOwner(makerOrder.Relayer, statedb)
 	matchingFee := big.NewInt(0)
 	// masternodes charges fee of both 2 relayers. If maker and Taker are on same relayer, that relayer is charged fee twice
-	matchingFee = matchingFee.Add(matchingFee, common.RelayerFee)
-	matchingFee = matchingFee.Add(matchingFee, common.RelayerFee)
+	matchingFee = matchingFee.Add(matchingFee, common.RelayerLendingFee)
+	matchingFee = matchingFee.Add(matchingFee, common.RelayerLendingFee)
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go:497]: charge relayerLendingFee only once (for borrowingRelayer only)

File: tomoxlending/order_processor.go [hunk 6]
```diff
@@ -400,203 +440,195 @@ func (l *Lending) getTradeQuantity(quoteInterest *big.Int, coinbase common.Addre
...
 	mapRelayerFee[takerOrder.Relayer] = newRelayerTakerFee
-	newRelayerMakerFee, err := lendingstate.CheckSubRelayerFee(makerOrder.Relayer, common.RelayerFee, statedb, mapRelayerFee)
+	newRelayerMakerFee, err := lendingstate.CheckSubRelayerFee(makerOrder.Relayer, common.RelayerLendingFee, statedb, mapRelayerFee)
 	if err != nil {
 		return err
 	}
 	mapRelayerFee[makerOrder.Relayer] = newRelayerMakerFee
-	lendingstate.SetSubRelayerFee(takerOrder.Relayer, newRelayerTakerFee, common.RelayerFee, statedb)
-	lendingstate.SetSubRelayerFee(makerOrder.Relayer, newRelayerMakerFee, common.RelayerFee, statedb)
+	lendingstate.SetSubRelayerFee(takerOrder.Relayer, newRelayerTakerFee, common.RelayerLendingFee, statedb)
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go]: charge borrowing fee
no investing fee

File: tomoxlending/order_processor.go [hunk 7]
```diff
@@ -610,9 +642,44 @@ func (l *Lending) ProcessCancelOrder(lendingStateDB *lendingstate.LendingStateDB
...
 		return nil, true
 	}
-	log.Debug("ProcessCancelOrder", "LendingToken", order.LendingToken, "CollateralToken", order.CollateralToken, "makerInterest", order.Interest, "lendingTokenDecimal", lendingTokenDecimal, "quantity", order.Quantity)
-	feeRate := lendingstate.GetExRelayerFee(order.Relayer, statedb)
-	tokenCancelFee := getCancelFee(lendingTokenDecimal, feeRate, order)
+	log.Debug("ProcessCancelOrder", "LendingToken", order.LendingToken, "CollateralToken", order.CollateralToken, "makerInterest", order.Interest, "lendTokenDecimal", lendTokenDecimal, "quantity", order.Quantity)
+	borrowFee := lendingstate.GetFee(statedb, order.Relayer)
+	collateralPrice := common.BasePrice
+
+	if order.Side == lendingstate.Borrowing {
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go:621]: similar to tomox trading, cancel orders include: lendToken, CollateralToken, term, lendId, hash
no order.Side
so please put tomochain/tomoxlending/order_processor.go:687 to the top of this function

File: tomoxlending/order_processor.go [hunk 8]
```diff
@@ -635,115 +702,113 @@ func (l *Lending) ProcessCancelOrder(lendingStateDB *lendingstate.LendingStateDB
 	return nil, false
 }
 
-func (l *Lending) ProcessDeposit(lendingStateDB *lendingstate.LendingStateDB, statedb *state.StateDB, chain consensus.ChainContext, coinbase common.Address, lendingOrderBook common.Hash, order *lendingstate.LendingItem) (error, bool) {
-	if err := lendingstate.CheckRelayerFee(order.Relayer, common.RelayerCancelFee, statedb); err != nil {
-		log.Debug("Relayer not enough fee when cancel order", "err", err)
+func (l *Lending) ProcessDeposit(lendingStateDB *lendingstate.LendingStateDB, statedb *state.StateDB, tradingStateDb *tradingstate.TradingStateDB, lendingOrderBook common.Hash, order *lendingstate.LendingItem) (error, bool) {
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go]: unused argument: `lendingOrderBook common.Hash`

File: tomoxlending/order_processor.go [hunk 9]
```diff
@@ -635,115 +702,113 @@ func (l *Lending) ProcessCancelOrder(lendingStateDB *lendingstate.LendingStateDB
...
-	}
-	masternodeOwner := statedb.GetOwner(coinbase)
-	statedb.AddBalance(masternodeOwner, common.RelayerCancelFee)
+	tradingStateDb.RemoveLiquidationPrice(tradingstate.GetTradingOrderBookHash(lendingTrade.CollateralToken, lendingTrade.LendingToken), lendingTrade.LiquidationPrice, lendingBook, lendingTrade.TradeId)
+
+	lendingstate.SubTokenBalance(order.UserAddress, order.Quantity, lendingTrade.CollateralToken, statedb)
+	lendingstate.AddTokenBalance(common.HexToAddress(common.LendingLockAddress), order.Quantity, lendingTrade.CollateralToken, statedb)
+	oldLockedAmount := lendingTrade.CollateralLockedAmount
+	newLockedAmount := new(big.Int).Add(order.Quantity, oldLockedAmount)
+	newLiquidationPrice := new(big.Int).Mul(lendingTrade.LiquidationPrice, oldLockedAmount)
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go:669]: newLiquidationPrice := new(big.Int).Mul(lendingTrade.LiquidationPrice, `newLockedAmount`)
newLiquidationPrice = new(big.Int).Div(newLiquidationPrice, oldLockedAmount)

**Review by thanhson1085 — APPROVED** [review]

**Pull request merged**